pkg/sampledriver/provisioner.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2021 The Kubernetes Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// You may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sampledriver
+
+import (
+	"context"
+
+	_ "k8s.io/klog/v2"
+	cosi "sigs.k8s.io/container-object-storage-interface-spec"
+)
+
+type ProvisionerServer struct {
+	provisioner string
+}
+
+func (s *ProvisionerServer) ProvisionerCreateBucket(ctx context.Context,
```

Comment:
When the provisioner creates the backend bucket but for some reason (provisioner crash, sidecar crash, etc.) the `ProvisionerCreateBucketResponse` is not handled, then the backend bucket is leaked. In order to solve this, I propose a two-step creation process:

1. `ProvisionerCreateBucketID` (gets called if `bucket.Spec.BucketID == ""`): The provisioner chooses a bucket ID and returns it in the response, but it doesn't create the backend bucket, yet. The sidecar persists the bucket ID to the `bucket.Spec.BucketID`.
2. `ProvisionerCreateBucket` (gets called if `bucket.Spec.BucketID != "" && !bucket.Status.BucketAvailable`): The provisioner now creates the backend bucket.

This way `ProvisionerCreateBucket` is retried with the same `BucketID` in case of abovementioned exceptional situation.

Reply:
> When the provisioner creates the backend bucket but for some reason (provisioner crash, sidecar crash, etc.) the ProvisionerCreateBucketResponse is not handled, then the backend bucket is leaked. In order to solve this, I propose a two-step creation process

This is not true. ProvisionerCreateBucket is idempotent. Bucket name + protocol is enough to ensure that.

Reply:
That is only the case if the bucket ID is deterministic. Are you saying that, for all provisioners, the bucket ID must be deterministically determined by the bucket name and protocol?